components/constituency_lookup/ConstituencyLookup.tsx

```diff
@@ -25,22 +34,93 @@ const errorCodeToErrorMessage = (code: ErrorCode) => {
   }
 };
 
+const lookupCache: {
+  [key: string]: Promise<ConstituencyLookupResponse | null>;
+} = {};
+
+const fetchApi = async (
+  requestBody: ConstituencyLookupRequest,
+): Promise<ConstituencyLookupResponse | null> => {
+  try {
+    console.log("Making API call to constituency lookup route");
+    const response = await fetch("/api/constituency_lookup", {
+      method: "POST",
+      body: JSON.stringify(requestBody),
+    });
+
+    if (response.ok) {
+      lookupCache[requestBody.postcode] = response.json();
+      return lookupCache[requestBody.postcode];
+    } else {
+      return null;
+    }
+  } catch {
+    return null;
+  }
+};
+
+type ReqControl = {
+  time: number;
+  timerID: ReturnType<typeof setTimeout> | null;
+  rateLimit: number;
+};
+
+const reqControl: ReqControl = {
+  time: 0,
+  timerID: null,
+  rateLimit: 3000,
+};
+
+// throttled apiFetch
+const throttledApi = async (
+  requestBody: ConstituencyLookupRequest,
+): Promise<ConstituencyLookupResponse | null> => {
+  if (lookupCache.hasOwnProperty(requestBody.postcode)) {
+    const cached = await lookupCache[requestBody.postcode];
+    if (cached !== null) {
+      return cached; //don't want to cache and return an error!
+    }
+  }
+
+  if (reqControl.timerID) {
+    clearTimeout(reqControl.timerID);
+  }
+
+  if (reqControl.time + reqControl.rateLimit < Date.now()) {
+    //more than rate limit ms since last request
+    reqControl.time = Date.now();
+    lookupCache[requestBody.postcode] = fetchApi(requestBody);
+  } else {
+    //need to delay the request
```

Comment:
Doesn't this end up meaning we still hit the API the same number of times (minus caching), just we defer multiple API calls which otherwise would've happened in quick succession?

An alternative might be that if the rate-limit has been hit, don't make the API call at all, and wait for the user to click the submit button before doing the lookup?

Reply:
The call to clearTimeout means you cancel previous requests that were issued within rateLimit time. So at most you can issue 1 request per rateLimit time period.

Reply:
Oh yeah, sorry, missed that line above :+1: